src/ripple/protocol/SystemParameters.h

```diff
@@ -37,23 +38,27 @@ systemName ()
 }
 
 /** Configure the native currency. */
-static
-std::uint64_t const
-SYSTEM_CURRENCY_GIFT = 1000;
 
+/** Number of drops in the genesis account. */
 static
-std::uint64_t const
-SYSTEM_CURRENCY_USERS = 100000000;
+constexpr
+XRPAmount
+INITIAL_XRP{ 100'000'000'000 * DROPS_PER_XRP };
 
-/** Number of drops per 1 XRP */
-static
-std::uint64_t const
-SYSTEM_CURRENCY_PARTS = 1000000;
+/** Returns true if the amount does not exceed the initial XRP in existence. */
+inline
+bool isLegalAmount (XRPAmount const& amount)
+{
+    return amount <= INITIAL_XRP;
+}
 
-/** Number of drops in the genesis account. */
-static
-std::uint64_t const
-SYSTEM_CURRENCY_START = SYSTEM_CURRENCY_GIFT * SYSTEM_CURRENCY_USERS * SYSTEM_CURRENCY_PARTS;
+inline
+bool isLegalAmount (std::uint64_t amount)
```

Comment:
I don't like this. It's a _dangerous_ API to use: it accepts a value in _drops_, making it really easy for someone to do the wrong thing. And it only gives us a marginal benefit in a couple of cases, in fee vote processing.

I'd prefer that, wherever this overload is called from, we just construct an instance of `XRPAmount` and pass that into the `isLegalAmount` variant above. That way it's very clear exactly what's happening and you can't accidentally pass in XRP where drops are expected, or vice-versa.

Reply:
AFAIK, outside of unit test code, every place that works with XRP uses a value expressed in drops, so I don't know how risky this really is. The bigger problem is that in the few places this overload is used, it's *also* dangerous to construct an `XRPAmount` directly because an absurdly large value will overflow `int64_t`, and thus end up as a negative value, which would also be really bad.

I'm really happy that I was able to convert existing fees to use `XRPAmount`, but there are a couple of cases where the signed/unsigned conversions make it just a little awkward.

There are a couple of potential solutions:
1. Get rid of the overload, check _only_ for `int64_t` overflow in these couple of places, then construct an `XRPAmount` as you suggested.
2. Define an unsigned `XRPAmount`-like instance of the `TaggedFee` class. and use that as the overload parameter for these couple of cases.
3. Decide that using `XRPAmount` wasn't the best idea and use a `TaggedFee` instance everywhere for fees.

I suspect @seelabs will have opinions about some of these options.